contracts/upgradeable_contracts/BridgeValidators.sol

```diff
@@ -9,89 +9,167 @@ import "../upgradeability/EternalStorage.sol";
+        require(_initialValidators.length == _initialRewards.length);
+
         for (uint256 i = 0; i < _initialValidators.length; i++) {
-            require(_initialValidators[i] != address(0));
-            assert(validators(_initialValidators[i]) != true);
+            require(_initialValidators[i] != address(0) && _initialValidators[i] != F_ADDR);
+            require(_initialRewards[i] != address(0));
+            require(!isValidator(_initialValidators[i]));
+
+            if (i == 0) {
+                setValidator(F_ADDR, _initialValidators[i]);
+                if (_initialValidators.length == 1) {
+                    setValidator(_initialValidators[i], F_ADDR);
+                }
+            } else if (i == _initialValidators.length - 1) {
+                setValidator(_initialValidators[i - 1], _initialValidators[i]);
+                setValidator(_initialValidators[i], F_ADDR);
+            } else {
+                setValidator(_initialValidators[i - 1], _initialValidators[i]);
+            }
+
             setValidatorCount(validatorCount().add(1));
-            setValidator(_initialValidators[i], true);
+            setValidatorRewardAddress(_initialValidators[i], _initialRewards[i]);
             emit ValidatorAdded(_initialValidators[i]);
         }
+
         uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
         uintStorage[keccak256("deployedAtBlock")] = block.number;
         setInitialize(true);
         emit RequiredSignaturesChanged(_requiredSignatures);
+
         return isInitialized();
     }
 
-    function addValidator(address _validator) external onlyOwner {
-        require(_validator != address(0));
+    function addValidator(address _validator, address _reward) external onlyOwner {
+        require(_validator != address(0) && _validator != F_ADDR);
+        require(_reward != address(0));
         require(!isValidator(_validator));
+
+        address firstValidator = getNextValidator(F_ADDR);
+        setValidator(_validator, firstValidator);
+        setValidatorRewardAddress(_validator, _reward);
+        setValidator(F_ADDR, _validator);
+
         setValidatorCount(validatorCount().add(1));
-        setValidator(_validator, true);
         emit ValidatorAdded(_validator);
     }
 
     function removeValidator(address _validator) external onlyOwner {
         require(validatorCount() > requiredSignatures());
         require(isValidator(_validator));
-        setValidator(_validator, false);
+        address validatorsNext = getNextValidator(_validator);
+        address index = F_ADDR;
+        address next = getNextValidator(index);
+
+        // find the element in the list pointing to _validator
+        while (next != _validator) {
+            index = next;
+            next = getNextValidator(index);
+        }
+
+        setValidator(index, validatorsNext);
+        deleteItemFromAddressStorage("validatorsList", _validator);
```

Comment:
could you confirm that after removing the mapping element is zeroed? Do we have the corresponding unit test for this?

Reply:
I added a unit test for this case. https://github.com/poanetwork/poa-bridge-contracts/pull/130/commits/349bbbb355d9ee155eefc53886780a4e6f3f7f4c